### PR TITLE
Expose `peakBitrate` property in `Format`

### PR DESCRIPTION
**Pull request by julioz**
Fixes https://github.com/google/ExoPlayer/issues/2863

Exposes `peakBitrate` value when both `BANDWIDTH` and `AVERAGE-BANDWIDTH` attributes are available.
The `bitrate` property will always carry the average bandwidth of the `Format` instance, while `peakBitrate`
will carry the value for the respective attribute, when present; according to
tools.ietf.org/html/draft-pantos-http-live-streaming-23#page-28

### CODE REVIEW

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylistParserTest.java [hunk 1]
```diff
@@ -236,14 +236,17 @@ public void testParseMasterPlaylist() throws IOException {
...
     HlsMasterPlaylist masterPlaylist =
         parseMasterPlaylist(PLAYLIST_URI, PLAYLIST_WITH_AVG_BANDWIDTH);
 
     List<HlsMasterPlaylist.Variant> variants = masterPlaylist.variants;
 
     assertThat(variants.get(0).format.bitrate).isEqualTo(1280000);
+    assertThat(variants.get(0).format.peakBitrate).isEqualTo(Format.NO_VALUE);
+
     assertThat(variants.get(1).format.bitrate).isEqualTo(1270000);
+    assertThat(variants.get(1).format.peakBitrate).isEqualTo(1280000);
```
Comment (julioz) [inline library/hls/src/test/java/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylistParserTest.java:249]: This is the example case for the presence of the attribute.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylistParserTest.java [hunk 2]
```diff
@@ -236,14 +236,17 @@ public void testParseMasterPlaylist() throws IOException {
   }
 
   @Test
-  public void testMasterPlaylistWithBandwdithAverage() throws IOException {
+  public void testMasterPlaylistWithBandwidthAverage() throws IOException {
```
Comment (julioz) [inline library/hls/src/test/java/com/google/android/exoplayer2/source/hls/playlist/HlsMasterPlaylistParserTest.java:239]: Fixed typo.

**Comment by julioz** [comment]
When running locally, some `Extractor` tests fail; I can't figure out why. I assume it's one of the (innumerous) `Format` constructors (or factory methods) that take in `bitrate` and didn't take in `peakBitrate`. Please advise before we merge this (or; feel free to fix and commit on top of this branch).

**Comment by icbaker** [comment]
Mind taking a look Santiago?

**Comment by AquilesCanta** [comment]
I think this is for @ojw28 to look at.

**Comment by AquilesCanta** [comment]
On my side, I think it's unfortunate we are using the `Format.bitrate` field for both `BANDWIDTH` and `AVERAGE-BANDWIDTH`. Adding the peak bitrate adds a bit of confusion to the mix.

Aside, I think it would be helpful for us to know why/how having both average and peak bitrates could be of use.

**Comment by julioz** [comment]
@AquilesCanta Please check the referred issue #2863 for context.

**Comment by AquilesCanta** [comment]
My bad, didn't see the link. However, the referenced issue makes the same question I made above. And an answer is still not available. It's worth looking into the issue first.

EDIT: @ojw28 removed the need more info tag from the issue, so I'm assuming he understands the use case. Thanks for your patience!

**Comment by tonihei** [comment]
I think the peak bitrate is preferable for adaptive track selection to be on the safe side, especially when the peak is much higher than the average. Otherwise, the ABR algorithm may run into issues during complex scenes.  

So, if available, we should probably just populate the `Format` bitrate by BANDWIDTH in all cases?

**Comment by julioz** [comment]
@tonihei That is for sure. The `BANDWIDTH` attribute is [_mandatory_](https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#page-28), while the `AVERAGE-BANDWIDTH` attribute is optional.
The change in https://github.com/google/ExoPlayer/commit/2795269daee2270fc2bda1512cefd86ef021f905 exposes the `AVERAGE-BANDWIDTH` value overriding the `BANDWIDTH` one, but that means we _hide away_ the latter. For advanced ABR algorithms that might need to introspect into that value, this becomes an issue; which this PR fixes.

Let me know if hiding one of the values is _by design_. If that's the case, we should document it publicly somewhere, I assume others might hit the same limitation in the future.

**Comment by AquilesCanta** [comment]
To provide some context: Back when AVERAGE-BANDWIDTH was added to the HLS spec, preferring AVERAGE-BANDWIDTH (which is option) over BANDWIDTH (which is mandatory) was a design decision. @tonihei, please feel free to take whichever approach you think is most suitable for ABR.

**Comment by ojw28** [comment]
> I think the peak bitrate is preferable for adaptive track selection to be on the safe side, especially when the peak is much higher than the average. Otherwise, the ABR algorithm may run into issues during complex scenes.

Presumably using both is preferable, else there wouldn't be value in them both existing. I agree that if we were to use just one, then peak bitrate is preferable, and so https://github.com/google/ExoPlayer/commit/53b52e784c2a7902b634348a2edfe81306318770 was probably a mistake.

I think we should expose both in `Format`, and use peak bitrate for ABR decisions for as long as we're only using one of them. I can sort this out whilst merging this pull request. I will separately propose a small fix to use peak bitrate ahead of merging the whole change.

**Comment by ojw28** [comment]
For what it's worth, the DASH specification gives a more nuanced definition of bitrate, that seems much more useful than either "peak" or "average". I guess it's closer to "peak" in practice:

> Consider a hypothetical constant bitrate channel of bandwidth with the value of this attribute in bits per second (bps). Then, if the Representation is continuously delivered at this bitrate, starting at any SAP that is indicated either by `@startWithSAP` or by any Segment Index box, a client can be assured of having enough data for continuous playout providing playout begins after `@minBufferTime * @bandwidth` bits have been received (i.e. at time `@minBufferTime` after the first bit is received).

Meanwhile the [HLS authoring specification](https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices) seems to allow you to specify average and peak bitrates that are wrong by up to 10%, whilst referencing another section that doesn't exist. Why?

> 1.26. For VOD content, the average segment bit rate MUST be within 10% of the AVERAGE-BANDWIDTH attribute. (See Declared versus measured values of bandwidths.)
> 1.27. For VOD content, the measured peak bit rate MUST be within 10% of the BANDWIDTH attribute.

**Comment by julioz** [comment]
@ojw28 @marcbaechinger It seems like we have opted for exposing the peak bitrate when available, super-seeding the avg bitrate data, which hinders more advanced ABR algorithms.
As this PR starts to get conflicts with `dev-v2`, I'm tempted to close it. As a side-note, it might be worth extracting some of the constructor calls with default values from many test suites into a test fixture class that can provide sensible values and still builders for further specialization if need be in tests. I didn't do it as part of this PR since I couldn't find other fixture classes in the codebase, but would gladly introduce this pattern as it has proven to be useful for us at SoundCloud.

Let me know what you think.

**Comment by ojw28** [comment]
Please leave it open. We'll resolve the conflict when we merge. Thanks!

**Comment by ojw28** [comment]
We need to make some bigger changes to land this, which we're working on now. These changes will land separately of this pull request, so I'll go ahead and close it. Please follow the tracking bug for updates. Thanks!

**Pull request closed without merging**